datafusion/core/src/physical_plan/wrapper.rs

```diff
@@ -0,0 +1,88 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+use arrow::row::{SortField, Rows};
+use arrow::datatypes::DataType;
+use arrow::error::ArrowError;
+use arrow_array::*;
+use arrow_array::types::*;
+use arrow::row::RowConverter;
+
+const LOW_CARDINALITY_THRESHOLD: usize = 10;
+
+macro_rules! downcast_dict {
+    ($array:ident, $key:ident) => {{
+        $array
+            .as_any()
+            .downcast_ref::<DictionaryArray<$key>>()
+            .unwrap()
+    }};
+}
+
+#[derive(Debug)]
+pub struct CardinalityAwareRowConverter {
+    fields: Vec<SortField>,
+    inner: Option<RowConverter>,
+    done: bool,
+}
+
+impl CardinalityAwareRowConverter {
+    pub fn new(fields: Vec<SortField>) -> Result<Self, ArrowError> {
+        Ok(Self {
+            fields: fields.clone(),
+            inner: None,
+            done: false,
+        })
+    }
+    
+    pub fn size(&self) -> usize {
+        return self.inner.as_ref().unwrap().size();
+    }
+
+    pub fn convert_rows(&self, rows: &Rows) -> Result<Vec<ArrayRef>, ArrowError> {
+        self.inner.as_ref().unwrap().convert_rows(rows)
+    }
+
+    pub fn convert_columns(
+        &mut self,
+        columns: &[ArrayRef]) -> Result<Rows, ArrowError> {
+        
+        if !self.done {
```

Comment:
If you made `fields: Option<Vec<SortField>>` you could drop self.done and this would become `if let Some(fields) = self.fields`

Alternatively an explicit enum might be better


```
pub struct CardinalityAwareRowConverter(ConverterState);

enum ConverterState {
    Init(Vec<SortField>),
    Ready(RowConverter)
}

```

Reply:
Umm , as per my understanding,`fields` is never None. We initialize `CardinalityAwareRowConverter` with some fields, and then modify it according to the first batch.

Reply:
Thanks for the pointer. Pushed the change.